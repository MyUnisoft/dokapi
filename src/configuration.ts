// Import Node.js Dependencies
import path from "path";
import fs from "fs/promises";

// Import Third-party Dependencies
import Ajv from "ajv";
import TOML from "@iarna/toml";

// Import Internal Dependencies
import schema from "./schema/configuration.schema.json";

// CONSTANTS
const kSchemaValidator = new Ajv();

export interface DokapiMenu {
  /** Title of the page (Will be extracted from markdown if not provided) */
  title?: string;

  /** System path to page or the directory containing the page */
  path: string;

  /** Is the menu collapsable (see stripe API for an example) */
  collapsable?: boolean;

  menu?: DokapiSubMenu[];
}

export type DokapiSubMenu = Omit<DokapiMenu, "menu" | "collapsable">;

export interface DokapiConfiguration {
  /** Title of the API */
  title: string;

  /** The logo to show under the title */
  logo?: string;

  /** The list of accessible menu (pages) on the left */
  menu: DokapiMenu[];
}

export type DokapiSubConfiguration = Pick<DokapiConfiguration, "title"> & {
  menu: DokapiSubMenu[]
};

/**
 * @description Read, parse and validate a Dokapi configuration.
 */
export async function read(location: string): Promise<DokapiConfiguration> {
  const str = await fs.readFile(
    path.join(location, "dokapi.config.toml"), "utf-8"
  );
  const manifest = TOML.parse(str);

  const validate = kSchemaValidator.compile(schema);
  if (!validate(manifest)) {
    console.log(validate.errors);

    throw new Error("Unable to validate Dokapi configuration");
  }

  return manifest as unknown as DokapiConfiguration;
}
