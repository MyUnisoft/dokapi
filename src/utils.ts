/**
 * @description Search for a title in a given Markdown string
 * @example
 * searchMarkdownHeadTitle("# hello world"); // hello world
 */
export function searchMarkdownHeadTitle(markdown: string): string | null {
  const result = /#(.*)/gm.exec(markdown);

  if (result !== null) {
    return result[1].trim();
  }

  return null;
}
