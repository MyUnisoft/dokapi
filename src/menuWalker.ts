// Import Internals
import * as DokapiConfig from "./configuration";

export async function* walkDokapiMenu(location: string, menuList: DokapiConfig.DokapiMenu[]) {
  for (const { title, path } of menuList) {
    console.log(title, path);

    yield "foobar";
  }
}
