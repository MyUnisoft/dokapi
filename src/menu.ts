// Import Node.js Dependencies
import path from "path";
import fs from "fs/promises";

// Import Third-party Dependencies
import MarkdownIt from "markdown-it";

// Import Internals
import * as DokapiConfig from "./configuration";
import * as utils from "./utils";

// VARS & CONSTANTS
const md = new MarkdownIt();

export type ParsedMenu = Pick<DokapiConfig.DokapiMenu, "title" | "collapsable"> & {
  /** HTML content (parsed from the menu markdown) */
  html: string;

  subMenu?: boolean;
};

export async function* walk(
  location: string,
  menuList: DokapiConfig.DokapiMenu[],
  subMenu = false
): AsyncIterableIterator<ParsedMenu> {
  for (const menu of menuList) {
    const menuPath = path.join(location, menu.path);
    const stat = await fs.stat(menuPath);

    const readmePath = stat.isDirectory() ? path.join(menuPath, "README.md") : menuPath;
    const readmeStr = await fs.readFile(readmePath, "utf-8");

    const title = menu.title ?? utils.searchMarkdownHeadTitle(readmeStr);
    if (typeof title === "undefined" || title === null) {
      throw new Error(`Unable to found a title for path: ${menuPath}`);
    }

    yield {
      title,
      collapsable: menu.collapsable ?? false,
      subMenu,
      html: md.render(readmeStr)
    };

    if (Array.isArray(menu.menu)) {
      yield* walk(stat.isDirectory() ? menuPath : path.dirname(menuPath), menu.menu, true);
    }
  }
}
