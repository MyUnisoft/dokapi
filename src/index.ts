// Import Node.js Dependencies
// import path from "path";
// import fs from "fs/promises";

// Import Internals
import * as DokapiConfig from "./configuration";
import { walkDokapiMenu } from "./menuWalker";

export interface generateDocumentationOptions {
  /** Default: `docs` */
  outDir?: string;
}

export async function generateDocumentation(location: string, options: generateDocumentationOptions = {}) {
  const { outDir = "docs" } = options;
  console.log(outDir);

  const config = await DokapiConfig.read(location);
  for await (const any of walkDokapiMenu(location, config.menu)) {
    console.log(any);
  }
}
