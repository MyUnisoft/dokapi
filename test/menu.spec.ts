/* eslint-disable arrow-body-style */

// Import Node.js Dependencies
import path from "path";

// Import Internal Dependencies
import * as DokapiConfig from "../src/configuration";
import * as DokapiMenu from "../src/menu";

// CONSTANTS
const kFixturesPath = path.join(__dirname, "fixtures");

describe("walk", () => {
  it("should be able to walk and parse Markdown files for the 'valid' fixture", async() => {
    const fixturePath = path.join(kFixturesPath, "valid");

    const { menu } = await DokapiConfig.read(fixturePath);
    const result: DokapiMenu.ParsedMenu[] = [];

    for await (const row of DokapiMenu.walk(fixturePath, menu)) {
      result.push(row);
    }

    const [first, second] = result;
    expect(first.title).toStrictEqual("Hello world");
    expect(first.collapsable).toBeFalsy();
    expect(first.subMenu).toBeFalsy();
    expect(first.html).toMatchSnapshot();

    expect(second.title).toStrictEqual("All products");
    expect(second.collapsable).toBeTruthy();
    expect(second.subMenu).toBeFalsy();
    expect(second.html).toMatchSnapshot();
  });

  it("should be able to walk sub menu", async() => {
    const fixturePath = path.join(kFixturesPath, "submenu");

    const menu: DokapiConfig.DokapiMenu[] = [
      {
        path: "main.md",
        menu: [
          { path: "subpage.md" }
        ]
      }
    ];
    const result: DokapiMenu.ParsedMenu[] = [];

    for await (const row of DokapiMenu.walk(fixturePath, menu)) {
      result.push(row);
    }

    const [first, second] = result;
    expect(first.title).toStrictEqual("Main page");
    expect(first.subMenu).toBeFalsy();

    expect(second.title).toStrictEqual("Sub page");
    expect(second.subMenu).toBeTruthy();
  });

  it("should throw if title is missing", async() => {
    expect.assertions(2);
    const fixturePath = path.join(kFixturesPath, "missingTitle");

    const menu: DokapiConfig.DokapiMenu[] = [
      { path: "empty.md" }
    ];

    try {
      const result: DokapiMenu.ParsedMenu[] = [];
      for await (const row of DokapiMenu.walk(fixturePath, menu)) {
        result.push(row);
      }
    }
    catch (error) {
      expect(error.name).toStrictEqual("Error");
      expect(error.message.includes("Unable to found a title for path")).toStrictEqual(true);
    }
  });
});
