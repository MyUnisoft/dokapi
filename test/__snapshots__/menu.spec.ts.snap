// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walk should be able to walk and parse Markdown files for the 'valid' fixture 1`] = `
"<h1>Introduction</h1>
<p>hello world</p>
"
`;

exports[`walk should be able to walk and parse Markdown files for the 'valid' fixture 2`] = `
"<h1>All products</h1>
<blockquote>
<p>Hey hey!</p>
</blockquote>
"
`;
